ngcc: add build marker

petebacondarwin opened the pull request:
This sits on top of #25406

petebacondarwin committed:
feat(ivy): implement ngcc build marker

`ngcc` adds marker files to each folder that has been
compiled, containing the version of the ngcc used.

When compiling, it will ignore folders that contain these
marker files, as long as the version matches.

mary-poppins commented:
You can preview 03b6121 at https://pr25557-03b6121.ngbuilds.io/.

gkalpak reviewed (commented):
AFAICT, more logic is needed. E.g. when we detect that a package has been compiled by a different ngcc version. 
I.e. we need a way to "uncompile" before compiling with the new version. ("Uncompiling" could be as simple as rimrafing `node_modules_ngtsc/` or as complex as walking `node_modules/` and renaming all backup files to their original name.)

I think we also need to record the compilations output path (e.g. `node_modules/` or `node_modules_ngtsc/`) in the marker metadata.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/transform/utils.ts:
```diff
@@ -6,12 +6,14 @@
 
 import {isDefined} from '../utils';
 
+export const NGCC_VERSION = '0.0.0-PLACEHOLDER';
```
I would rather not have a dependency on compiler-cli if we can avoid it. That version is created in exactly the same way as this anyway.

gkalpak replied:
Hm...I thought `ngcc` will be part of the `compiler-cli` package anyway.
In any case, not a big deal :grin:

petebacondarwin replied:
To give more strength to this approach I can also say that I conferred with Alex Eagle about the best approach here as well :-)

mary-poppins commented:
You can preview a616204 at https://pr25557-a616204.ngbuilds.io/.

petebacondarwin commented:
We should avoid trying to "uncompile" at this stage. The backups are there if we need them for adding such a strategy later. In the meantime I will add a message telling the user to blow away their node_modules if the ngcc version changes.

petebacondarwin committed:
fixup! feat(ivy): implement ngcc build marker

mary-poppins commented:
You can preview a35ec67 at https://pr25557-a35ec67.ngbuilds.io/.

mary-poppins commented:
You can preview 02c2f33 at https://pr25557-02c2f33.ngbuilds.io/.

mary-poppins commented:
You can preview a6e05a5 at https://pr25557-a6e05a5.ngbuilds.io/.

petebacondarwin commented:
We just need to work out what to do about the failing integration test: bazel tests do not allow the subject to write over the node_modules of the test folder.

petebacondarwin committed:
test(ivy): mock the filesystem in ngcc main() integration tests

Bazel does not like the filesystem being modified.
This commit a temporary mock filesystem that can be modified as needed.

mary-poppins commented:
You can preview 137a198 at https://pr25557-137a198.ngbuilds.io/.

mary-poppins commented:
You can preview 736507f at https://pr25557-736507f.ngbuilds.io/.

petebacondarwin commented:
I have implemented an (in-memory) mock filesystem for the integration test that was failing.
This allows the test to be able to write to the `node_modules` folder safely. It could also allow
us to do more subtle testing of file changes at this level, going forward.

petebacondarwin commented:
Awaiting G3 resubmit and pullapprove for `integration` test.
^^^ @alxhub or @mhevery
